### DISCUSSION TITLE
Markdown alignment

### DISCUSSION THREAD

**Discussion by sandstrom** [typst/typst#2498] [Ideas]
To drive typst adoption, I think there is an opportunity to align the syntax closer to markdown. Not 100% alignment though, we shouldn't adopt markdown concepts when they don't make sense!

But being closer to markdown would make it easier for people to convert documents (not everyone can script it), and make the mental step lower.

For example, a `## Heading` in markdown looks like this. In typst it's `== Heading`. Why not stick to the markdown convention?

I know `#` currently used for function invocations, but maybe there is another character we could use for that instead?

Not a major thing, I think typst will be successful regardless. But before 1.0 there is an opportunity to refine the syntax, and this is something I'd consider.

#### Aside

As an aside, make sure you reserve some unused ascii characters for future use. For example, if `~` or `<` doesn't have a special meaning yet. Don't let it be a "regular character", reserve it for future use instead. I'm 100% certain you'll thank yourself in the future! (and avoid breaking changes when introducing new syntax in typst 2.x)

**Comment by Dherse** [comment]
No.

**Reply by sandstrom** [reply]
That was a quick shoot down! 😄 

![image](https://github.com/typst/typst/assets/122287/b2ee28bb-22dd-41a9-9a0f-3f753b6d0508)

**Reply by Dherse** [reply]
Sorry 😂

**Reply by jamesrswift** [reply]
I get that you might not like the idea, but you could at least be constructive in your dismissal.

**Reply by Dherse** [reply]
Ok, here is a list in no particular order:

- Typst is not trying to be a replacement for Markdown
- Typst is not trying to emulate Markdown's syntax, any similarity is purely coincidental
- Markdown targets a different set of use cases (although some use cases are in common)
- Typst is not Markdown
- This would make the syntax more confusing as it would introduce multiple ways of doing the same thing
- Markdown is terrible and Typst is not
- Typst is great and Markdown is terrible
- It would make the parser and the AST unnecessarily complex
- Adoption does not require emulation otherwise we would be using `\` everywhere
- Pandoc exists for a reason

**Reply by laurmaedje** [reply]
> ... any similarity is purely coincidental

That's not entirely true. There is similarly and it's there for familiarity to make it easier to learn. However, in the case of `=`, I decided that it's worth it to change this because the `#` is very important for Typst and it's the best character for this in my opinion. There is also precedent for `=` for headings with AsciiDoc.

Beyond that Typst skips some of Markdown's syntax because it's less used and we can just express it as a function, keeping the syntax only for the most important thing. In the end, Typst is quite different from Markdown.

**Comment by Enivex** [comment]
You can already use pandoc to convert markdown to typst

**Comment by SebastianJL** [comment]
@Dherse 
> * This would make the syntax more confusing as it would introduce multiple ways of doing the same thing

Not necessarily. You could in principle just exchange `=` and `#`. `=` For example has a precedence as a formula sign in excel. Not entirely awful.

> * It would make the parser and the AST unnecessarily complex

Again, if you just exchange `#` and `=` I don't see why this would be the case.

> * Pandoc exists for a reason

Fair enough.

@laurmaedje 
> Beyond that Typst skips some of Markdown's syntax because it's less used and we can just express it as a function, keeping the syntax only for the most important thing. In the end, Typst is quite different from Markdown.

I fully agree that it's the best course of action to only put a few select things into syntax instead of functions. However, that is not really an argument against this singular change. After all, headings already have a special case. And maybe I'm wrong but I think markdown is better known than asciidoc these days. And I mean the very basic syntax, headers `#`, lists `-` and stuff like `` `monospace` `` and `` ```code env``` ``.

So maybe killing your darlings could make sense here?

**Reply by laurmaedje** [reply]
I think `# Heading` is okay, but `=func(..)` just looks bad. Meanwhile `= Heading` actually makes a lot of sense if you think about (similar to `====== Section ======` which is a natural division). And `#func(..)` also looks nice. This is not something we will change at this point.

**Reply by Dherse** [reply]
I would also say that to be a superset of Markdown would involve supporting all of Markdown's syntax 🤷‍♂️

**Reply by SebastianJL** [reply]
Yeah I think this is either a misunderstanding, or OG means something different than I. But at least they write 

> To drive typst adoption, I think there is an opportunity to align the syntax closer to markdown. Not a 100%, we shouldn't adopt their concepts when they don't make sense (it shouldn't be a superset in the strict sense).

I think it's more of an opportunity to align the two a bit more. The word super-set is a sub-optimal choice in my opinion.

Anyways, my point was to consider it, not that it would be strictly better nor that the current solution isn't satisfactory.

**Reply by sandstrom** [reply]
Correct! I've updated the title to clarify.

To add some more thoughts, I think Typst will be successful regardless.

But Markdown is popular and well known, so if I were you I would do an honest review of syntax sometime before 1.0, and see if there are opportunities for alignment.

Such review shouldn't only be about syntax, also ask people what other things they think are confusing or complicated. No one gets the design perfect at the first attempt, and changing is easier before 1.0 than after.
